### Import zbioru danych

In [ ]:
import pandas as pd

df = pd.read_csv('AmesHousing.csv')

### Rozmiar zbioru danych

In [24]:
df.shape

(2930, 82)

Nasz `unikalny` zbior danych sklada sie z `2930` obserwacji oraz `82` cech.

### Unikatowe Wartości

In [38]:
uniqueCounts = df.nunique().sort_values(ascending=True)

uniqueCounts_dataFrame = uniqueCounts.reset_index()
uniqueCounts_dataFrame.columns = ['Feature', 'Unique Values']
uniqueCounts_dataFrame
# print(uniqueCounts_dataFrame.to_string(index=False))

,Feature,Unique Values
0,Central Air,2
1,Street,2
2,Alley,2
3,Bsmt Half Bath,3
4,Paved Drive,3
...,...,...
77,Bsmt Unf SF,1137
78,Gr Liv Area,1292
79,Lot Area,1960
80,PID,2930


Powyższa tabela składa się z trzech kolumn:

1. **Index**  
   Automatycznie wygenerowany numer wiersza

2. **Feature**  
   Nazwa cechy (kolumny) w zbiorze danych.

3. **Unique Values**  
   Liczba unikatowych wartości przyjmowanych przez daną cechę w całym zbiorze obserwacji.  
---
- Jeśli **Unique Values = 2**, to zmienna jest binarna (np. 0/1 lub TAK/NIE).  
  – Przykład: `Central Air` przyjmuje wartości `Y`/`N`, stąd ma 2 unikalne wartości.

- **Cechy dyskretne** to te o niewielkiej liczbie unikatów (zwykle 3–10).  
  – Przykłady: `Fireplace` przyjmuje juz 5 wartosci

- **Cechy ciągłe** mają bardzo dużo unikatowych wartości,   
  – Przykłady: `Lot Area` przyjmuje 1960 unikalnych wartosci

Dzięki temu podziałowi łatwo zidentyfikujemy, które cechy traktujemy jako dyskretne, a które jako ciągłe i z nich wybierzemy dwa zestawy po 7 cech dla kazdego typu


### Selekcja cech dyskretnych oraz ciągłych

In [36]:
discreteFeatures = [
    'Central Air',
    'Paved Drive',
    'Bsmt Full Bath',
    'Kitchen AbvGr',
    'Fireplaces',
    'Bsmt Half Bath',
    'Garage Cars'
]

1. **`Central Air`**  
   - **Opis:** Czy budynek posiada klimatyzację (`Y`/`N`).  
   - **Typ:** Zmienna binarna (2 wartości).  
   - **Zakres wartości:** `Y` (tak), `N` (nie).  

2. **`Paved Drive`**  
   - **Opis:** Rodzaj utwardzenia podjazdu.  
   - **Typ:** Zmienna nominalna.  
   - **Zakres wartości:** `N` (brak utwardzenia), `P` (częściowe), `Y` (tak).  

3. **`Bsmt Full Bath`**  
   - **Opis:** Liczba pełnych łazienek w piwnicy.  
   - **Typ:** Zmienna licznikowa.  
   - **Zakres wartości:** 0–3.  

4. **`Kitchen AbvGr`**  
   - **Opis:** Liczba kuchni nad poziomem gruntu.  
   - **Typ:** Zmienna licznikowa.  
   - **Zakres wartości:** zwykle 1–2.  

5. **`Fireplaces`**  
   - **Opis:** Liczba kominków w domu.  
   - **Typ:** Zmienna licznikowa.  
   - **Zakres wartości:** 0–4.  

6. **`Bsmt Half Bath`**  
   - **Opis:** Liczba „połówkowych” łazienek w piwnicy.  
   - **Typ:** Zmienna licznikowa.  
   - **Zakres wartości:** 0–2.  

7. **`Garage Cars`**  
   - **Opis:** Liczba miejsc parkingowych w garażu.  
   - **Typ:** Zmienna licznikowa.  
   - **Zakres wartości:** 0–4.  

---

*Zdecydowalismy sie na wybor takich cech, poniewaz kazda z nich ma liczbe unikatowych wartosci (<10), dobrze bedzie sie je wizualizowac, uwazamy ze dobrze oddadza wplyw na atrakcyjnosc mieszkania*


In [ ]:
continuousFeatures = [
    'Lot Area',
    'Gr Liv Area',
    'Total Bsmt SF',
    '1st Flr SF',
    '2nd Flr SF',
    'Garage Area',
    'SalePrice'
]

1. **`Lot Area`**  
   - **Opis:** Powierzchnia działki w stopach kwadratowych.  
   - **Typ:** Zmienna ciągła.  
   - **Zakres wartości:** od kilku tysięcy do ponad 200 000 ft².

2. **`Gr Liv Area`**  
   - **Opis:** Powierzchnia użytkowa mieszkalna powyżej poziomu gruntu w stopach kwadratowych. 
   - **Typ:** Zmienna ciągła.  
   - **Zakres wartości:** od kilkuset do ponad 4500 ft².

3. **`Total Bsmt SF`**  
   - **Opis:** Całkowita powierzchnia piwnicy w stopach kwadratowych.  
   - **Typ:** Zmienna ciągła.  
   - **Zakres wartości:** od 0 (brak piwnicy) do około 6000 w stopach kwadratowych.

4. **`1st Flr SF`**  
   - **Opis:** Powierzchnia parteru w stopach kwadratowych.  
   - **Typ:** Zmienna ciągła.  
   - **Zakres wartości:** od kilkuset do ponad 4000 ft².

5. **`2nd Flr SF`**  
   - **Opis:** Powierzchnia piętra w stopach kwadratowych. 
   - **Typ:** Zmienna ciągła.  
   - **Zakres wartości:** od 0 (domy parterowe) do około 2000 ft².

6. **`Garage Area`**  
   - **Opis:** Powierzchnia garażu w stopach kwadratowych. 
   - **Typ:** Zmienna ciągła.  
   - **Zakres wartości:** od 0 (brak garażu) do ponad 1400 ft².

7. **`SalePrice`**  
   - **Opis:** Cena sprzedaży domu w dolarach amerykańskich.  
   - **Typ:** Zmienna ciągła.  
   - **Zakres wartości:** od kilkudziesięciu tysięcy do ponad miliona dolarów.


In [42]:
myData = discreteFeatures + continuousFeatures
summary = pd.DataFrame({
    'Feature': myData,
    'Type':   ['discrete']*len(discreteFeatures) + ['continuous']*len(continuousFeatures),
    'Unique Values': [df[f].nunique() for f in myData]
})
summary

,Feature,Type,Unique Values
0,Central Air,discrete,2
1,Paved Drive,discrete,3
2,Bsmt Full Bath,discrete,4
3,Kitchen AbvGr,discrete,4
4,Fireplaces,discrete,5
5,Bsmt Half Bath,discrete,3
6,Garage Cars,discrete,6
7,Lot Area,continuous,1960
8,Gr Liv Area,continuous,1292
9,Total Bsmt SF,continuous,1058


Nasz nowy zbiór, na którym będziemy pracować, składa się z:
- **14 cech**:  
  - 7 cech dyskretnych  
  - 7 cech ciągłych  
- **2930 obserwacji**  

Ograniczylismy sie do tych 14 cech, zatem duzo latwiejsze staje sie przeprowadzenie analizy dancyh jak i sama wizualizacja danych.  
Dodatkwo zachowujemy balans miedzy zmiennyli liczbowymi a kategorycznymi.


### Opis zbioru danych

Zdecydowaliśmy sie na **Ames Housing Dataset** jako nasz unikalny zbiór danych.  
Zbiór danych pochodzi z Kaggle [Ames Data Set](https://www.kaggle.com/datasets/shashanknecrothapa/ames-housing-dataset).  

- **Liczba obserwacji:** 2 930  
- **Liczba cech:** 79  
- **Cechy:** różnorodne atrybuty opisujące domy  
- **Zmienna docelowa:** `SalePrice` (cena sprzedaży domu w USD)  

Opis zbioru danych na Kaggle:
> _“The Ames Housing Dataset is widely used in the machine learning community for tasks such as regression modeling, feature engineering, and predictive analytics related to housing prices. It serves as a valuable resource for developing and testing machine learning algorithms and techniques in the real estate domain.”_    

Podsumowujac, z tychze 79 cech wybralismy:
- ogolne 14 cech
    - 7 cech dyskretnych  
    - 7 cech ciągłych


Dzięki wybraniu 7 zmiennych dyskretnych i 7 zmiennych ciągłych, uzyskujemy zrównoważony zestaw danych idealny do dalszej analizy statystycznej.


TODO:
- wizualizacja kazdej cechy ???


### Estymacja punktowa

#### Plan analizy cech dyskretnych

**Cechny licznikowe**
  1. Średnia (mean)  
  2. Wariancja (variance)  
  3. Odchylenie standardowe (std)  
  4. Odchylenie przeciętne (MAD)  
  5. Mediana (median)  
  6. Moda (mode)  
  7. Skośność (skewness)  
  8. Kurtoza (kurtosis)  
  9. 1-szy i 3-ci kwartyl (quantiles 0.25, 0.75)  
  10. Rozstęp międzykwartylowy (IQR = Q3–Q1)

**Cechny nominalne**
  1. Tabelę częstości (count)  
  2. Udział (proportion)  
  3. Dominującą kategorię (mode)

In [6]:
import pandas as pd
import numpy as np
from scipy import stats

myData = pd.read_csv('AmesHousing.csv')

continuousFeatures_count = [
    'Bsmt Full Bath',
    'Kitchen AbvGr',
    'Fireplaces',
    'Bsmt Half Bath',
    'Garage Cars'
]
continuousFeatures_nominal = [
    'Central Air',
    'Paved Drive'
]

results = []
for f in continuousFeatures_count:
    x = myData[f].dropna()
    mu    = x.mean()
    var   = x.var(ddof=1)
    std   = x.std(ddof=1)
    mad   = (x - mu).abs().mean()
    med   = x.median()
    mode  = x.mode().iloc[0]
    skew  = x.skew()
    kurt  = x.kurtosis()
    q1, q3 = x.quantile([0.25, 0.75])
    iqr   = q3 - q1

    results.append({
        'Feature':       f,
        'Mean':          mu,
        'Variance':      var,
        'Std':       std,
        'MAD':           mad,
        'Median':        med,
        'Mode':          mode,
        'Skewness':      skew,
        'Kurtosis':      kurt,
        'Q1 (25%)':      q1,
        'Q3 (75%)':      q3,
        'IQR':           iqr
    })

continuousFeatures_countStats = pd.DataFrame(results).set_index('Feature')
display(continuousFeatures_countStats)

,Mean,Variance,Std,MAD,Median,Mode,Skewness,Kurtosis,Q1 (25%),Q3 (75%),IQR
Feature,,,,,,,,,,,
Bsmt Full Bath,0.431352,0.275436,0.524820,0.502950,0.0,0.0,0.616639,-0.747989,0.0,1.0,1.0
Kitchen AbvGr,1.044369,0.045829,0.214076,0.086818,1.0,1.0,4.313825,19.869743,1.0,1.0,0.0
Fireplaces,0.599317,0.419802,0.647921,0.581727,1.0,0.0,0.739215,0.101508,0.0,1.0,1.0
Bsmt Half Bath,0.061134,0.060149,0.245254,0.114960,0.0,0.0,3.940795,14.921740,0.0,0.0,0.0
Garage Cars,1.766815,0.578461,0.760566,0.596771,2.0,2.0,-0.219836,0.244969,1.0,2.0,1.0


Teraz krotko zinterpretujemy otrzymane wyniki, opiszemy krotko kazdy estymator o czym mówi.

`Średnia` - to suma wartosci podzielona przez liczbe obserwacji.  
- W naszym przypadku widzimy, w wiekszosci domom jest mnie niz pol pelnej lazienki w piwnicy, wiec wiekszosc domow jej nie ma.  
- Niemal zawsze jest jedna kuchnia (nad poziomem gruntu), czasem nawet dwie sa kuchnie.  
- Wiekszosc domow ma zazwyczaj jeden kominek albo wcale.  
- Rzadko wystepuja polowy lazienek w piwnicy
- srednio domy maja prawie 2 miesjca prawkingowe.

`Wariancja` - to miara rozrzutu wokol sredniej
- **Bsmt Full Bath:** 0.28 – umiarkowany rozrzut wokół 0.43.  
- **Kitchen AbvGr:** 0.05 – bardzo mały rozrzut, prawie każda wartość to 1.  
- **Fireplaces:** 0.42 – umiarkowany rozrzut (0 lub 1, rzadziej 2–3).  
- **Bsmt Half Bath:** 0.06 – rozrzut praktycznie zerowy (prawie wszędzie 0).  
- **Garage Cars:** 0.58 – umiarkowany rozrzut wokół ~1.8. 

`Odchylenie standardowe` - pierwiastek z wariancji, srednie odchylenie od sredniej
- **Bsmt Full Bath:** 0.52  
- **Kitchen AbvGr:** 0.21  
- **Fireplaces:** 0.65  
- **Bsmt Half Bath:** 0.25  
- **Garage Cars:** 0.76 

`Odchylenie przecietne (MAD)` - Średnia z bezwzględnych odchyleń od średniej, jest mniej wrazliwa na wartosci odstajace niz odch. std.
- **Bsmt Full Bath:** 0.50 – przeciętnie odchylenie od średniej to pół łazienki.  
- **Kitchen AbvGr:** 0.09 – większość domów ma dokładnie jedną kuchnię.  
- **Fireplaces:** 0.58  
- **Bsmt Half Bath:** 0.11  
- **Garage Cars:** 0.60 

`Mediana` - Wartość środkowa w uporządkowanym zbiorze. 
- **Bsmt Full Bath:** 0 – co najmniej połowa domów nie ma pełnej łazienki w piwnicy.  
- **Kitchen AbvGr:** 1 – przynajmniej połowa ma jedną kuchnię.  
- **Fireplaces:** 1 – przynajmniej połowa ma co najmniej 1 kominek.  
- **Bsmt Half Bath:** 0 – połowa domów nie ma połowy łazienki w piwnicy.  
- **Garage Cars:** 2 – połowa garaży pomieści co najmniej dwa auta.

`Moda` - Najczęściej występująca wartość.   
- **Bsmt Full Bath:** 0 – najwięcej domów nie ma pełnej łazienki w piwnicy.  
- **Kitchen AbvGr:** 1 – przeważnie jedna kuchnia.  
- **Fireplaces:** 0 – najczęściej brak kominka.  
- **Bsmt Half Bath:** 0 – najczęściej brak połówkowej łazienki w piwnicy.  
- **Garage Cars:** 2 – najczęściej garaż na dwa samochody.

`Kwantyle` - 
- **Q1 (25%)** – wartość poniżej której znajduje się 25% danych.  
- **Q3 (75%)** – wartość poniżej której znajduje się 75% danych.  
- **IQR = Q3 – Q1** – rozstęp środkowych 50% danych.  
  | feature            | Q1  | Q3  | IQR |
  |--------------------|-----|-----|-----|
  | Bsmt Full Bath     | 0   | 1   | 1   |
  | Kitchen AbvGr      | 1   | 1   | 0   |
  | Fireplaces         | 0   | 1   | 1   |
  | Bsmt Half Bath     | 0   | 0   | 0   |
  | Garage Cars        | 1   | 2   | 1   |

`Skośność` - mierzy asyetrie rozkladu wzgledem sredniej, >0 prawy ogon, <0 lewy ogon 
- **Bsmt Full Bath:** 0.62 – (kilka domów ma więcej niż 1 łazienkę).  
- **Kitchen AbvGr:** 0.31 – (nieliczni mają 2 kuchnie).  
- **Fireplaces:** 0.74 – umiarkowana skośność dodatnia.  
- **Bsmt Half Bath:** 3.94 – (praktycznie wszędzie 0, nieliczne 1–2).  
- **Garage Cars:** –0.22 – (kilka domów ma 0 lub 1 miejsce, większość 2).

`Kurtoza` - >0 mowi ze dane ukaldaja sie w ostre szczyty, <0 rozkald jest splaszczony 
- **Bsmt Full Bath:** –0.75 – lekko spłaszczony rozkład.  
- **Kitchen AbvGr:** 19.87 – ekstremalnie wysokie skupienie na 1 i rzadkie „wpadki” na 2 (ciężkie ogony).  
- **Fireplaces:** 0.10 – nieznacznie cięższe ogony.  
- **Bsmt Half Bath:** 14.92 – bardzo ciężkie ogony (zbiór prawie wyłącznie z zerami).  
- **Garage Cars:** 0.24 – lekkie wydłużenie ogonów.
 


In [ ]:
import pandas as pd

for feat in continuousFeatures_nominal:
    counts = myData[feat].value_counts(dropna=False)
    props  = myData[feat].value_counts(normalize=True, dropna=False)
    freq_df = pd.DataFrame({
        'category':   counts.index,
        'count':      counts.values,
        'proportion': props.values
    })
    display(freq_df)



,category,count,proportion
0,Y,2734,0.933106
1,N,196,0.066894


,category,count,proportion
0,Y,2652,0.905119
1,N,216,0.073720
2,P,62,0.021160


`Central Air`
- **Y (tak):** 2 734 obserwacje (~93%)  
- **N (nie):** 196 obserwacji (~7%)   
> Większość domów w Ames wyposażona jest w centralną klimatyzację.

`Paved Drive ` 
- **Y (tak):** 2 652 obserwacje (91%)  
- **N (brak):** 216 obserwacji (7%)  
- **P (częściowo):**  62 obserwacje (2%)  
> Ponad 90% posesji ma w pełni utwardzony podjazd, a jedynie niewielki odsetek jest utwardzony częściowo.  
